Index: linux-2.6.36/include/linux/kernel.h
===================================================================
--- linux-2.6.36.orig/include/linux/kernel.h
+++ linux-2.6.36/include/linux/kernel.h
@@ -756,7 +756,9 @@ struct sysinfo {
 };
 
 /* Force a compilation error if condition is true */
+#ifndef BUILD_BUG_ON
 #define BUILD_BUG_ON(condition) ((void)BUILD_BUG_ON_ZERO(condition))
+#endif
 
 /* Force a compilation error if condition is constant and true */
 #define MAYBE_BUILD_BUG_ON(cond) ((void)sizeof(char[1 - 2 * !!(cond)]))
Index: linux-2.6.36/net/ipv4/netfilter/ip_tables.c
===================================================================
--- linux-2.6.36.orig/net/ipv4/netfilter/ip_tables.c
+++ linux-2.6.36/net/ipv4/netfilter/ip_tables.c
@@ -40,6 +40,7 @@ MODULE_DESCRIPTION("IPv4 packet filter")
 
 #if defined (CONFIG_NAT_FCONE) || defined (CONFIG_NAT_RCONE)
 unsigned char wan_name[IFNAMSIZ];
+EXPORT_SYMBOL(wan_name);
 #endif
 
 #ifdef DEBUG_IP_FIREWALL
