Index: linux-2.6.36/sound/core/Kconfig
===================================================================
--- linux-2.6.36.orig/sound/core/Kconfig
+++ linux-2.6.36/sound/core/Kconfig
@@ -8,7 +8,7 @@ config SND_PCM
 	select GCD
 
 config SND_HWDEP
-	tristate
+	tristate "Sound hardware support"
 
 config SND_RAWMIDI
 	tristate
Index: linux-2.6.36/arch/mips/ralink/setup.c
===================================================================
--- linux-2.6.36.orig/arch/mips/ralink/setup.c
+++ linux-2.6.36/arch/mips/ralink/setup.c
@@ -77,6 +77,8 @@ extern void mips_time_init(void);
 extern void mips_timer_setup(struct irqaction *irq);
 
 int coherentio = -1;    /* no DMA cache coherency (may be set by user) */
+EXPORT_SYMBOL(coherentio);
+
 int hw_coherentio;      /* init to 0 => no HW DMA cache coherency (reflects real HW) */
 static int __init setcoherentio(char *str)
 {
