Index: linux-2.6.36/drivers/net/pptp.c
===================================================================
--- linux-2.6.36.orig/drivers/net/pptp.c
+++ linux-2.6.36/drivers/net/pptp.c
@@ -1186,7 +1186,7 @@ static struct pppox_proto pppox_pptp_pro
 #endif
 };
 
-#if defined(CONFIG_GRE) || defined(CONFIG_GRE_MODULE)
+#if 1 // defined(CONFIG_GRE) || defined(CONFIG_GRE_MODULE)
 static struct gre_protocol gre_pptp_protocol = {
 	.handler	= pptp_rcv,
 };
@@ -1220,7 +1220,7 @@ static int __init pptp_init_module(void)
 		return -ENOMEM;
 	}
 
-#if defined(CONFIG_GRE) || defined(CONFIG_GRE_MODULE)
+#if 1 // defined(CONFIG_GRE) || defined(CONFIG_GRE_MODULE)
 	if (gre_add_protocol(&gre_pptp_protocol, GREPROTO_PPTP) < 0) {
 		printk(KERN_INFO "PPTP: can't add protocol\n");
 		goto out_free_mem;
@@ -1258,7 +1258,7 @@ out_unregister_sk_proto:
 out_inet_del_protocol:
 #endif
 
-#if defined(CONFIG_GRE) || defined(CONFIG_GRE_MODULE)
+#if 1 // defined(CONFIG_GRE) || defined(CONFIG_GRE_MODULE)
 	gre_del_protocol(&gre_pptp_protocol, GREPROTO_PPTP);
 #elif LINUX_VERSION_CODE < KERNEL_VERSION(2,6,0)
 	inet_del_protocol(&net_pptp_protocol);
@@ -1274,7 +1274,7 @@ out_free_mem:
 static void __exit pptp_exit_module(void)
 {
 	unregister_pppox_proto(PX_PROTO_PPTP);
-#if defined(CONFIG_GRE) || defined(CONFIG_GRE_MODULE)
+#if 1 // defined(CONFIG_GRE) || defined(CONFIG_GRE_MODULE)
 #if LINUX_VERSION_CODE > KERNEL_VERSION(2,6,0)
 	proto_unregister(&pptp_sk_proto);
 #endif
Index: linux-2.6.36/drivers/net/Kconfig
===================================================================
--- linux-2.6.36.orig/drivers/net/Kconfig
+++ linux-2.6.36/drivers/net/Kconfig
@@ -3324,6 +3324,17 @@ config PPP_MPPE
 	 See http://pptpclient.sourceforge.net/ for information on
 	 configuring PPTP clients and servers to utilize this method.
 
+config PPTP
+        tristate "PPP over IPv4 (PPTP)"
+        depends on PPP && NET_IPGRE_DEMUX
+        ---help---
+          Support for PPP over IPv4.(Point-to-Point Tunneling Protocol)
+
+          This driver requires pppd plugin to work in client mode or
+          modified pptpd (poptop) to work in server mode.
+          See http://accel-pptp.sourceforge.net/ for information how to
+          utilize this module.
+
 config PPPOE
 	tristate "PPP over Ethernet (EXPERIMENTAL)"
 	depends on EXPERIMENTAL && PPP
Index: linux-2.6.36/crypto/Kconfig
===================================================================
--- linux-2.6.36.orig/crypto/Kconfig
+++ linux-2.6.36/crypto/Kconfig
@@ -34,11 +34,13 @@ config CRYPTO_FIPS
 config CRYPTO_ALGAPI
 	tristate
 	select CRYPTO_ALGAPI2
+	default y
 	help
 	  This option provides the API for cryptographic algorithms.
 
 config CRYPTO_ALGAPI2
 	tristate
+	default y
 
 config CRYPTO_AEAD
 	tristate
